src/ovs/aca_ovs_l3_programmer.cpp

```diff
@@ -42,49 +42,346 @@ ACA_OVS_L3_Programmer &ACA_OVS_L3_Programmer::get_instance()
+        // since message_type is full, we can simply remove the existing entry and use
+        // the next router entry population block
+        overall_rc = delete_router(current_RouterConfiguration, dataplane_programming_time);
+        if (overall_rc != EXIT_SUCCESS) {
+          throw std::runtime_error("Failed to delete an existing router entry");
+        }
+      } else {
+        // current_RouterConfiguration.message_type() == MessageType::DELTA
+        // carefully update the existing entry with new information
+        new_subnet_routing_tables = _routers_table[router_id];
+      }
+    }
+
+    // ==============================
+    // router entry population block
+    // ==============================
+
+    // it is okay for have subnet_routing_tables_size = 0
+    for (int i = 0; i < current_RouterConfiguration.subnet_routing_tables_size(); i++) {
+      auto current_subnet_routing_table =
+              current_RouterConfiguration.subnet_routing_tables(i);
+
+      string current_router_subnet_id = current_subnet_routing_table.subnet_id();
+
+      ACA_LOG_DEBUG("Processing subnet ID: %s for router ID: %s.\n",
+                    current_router_subnet_id.c_str(),
+                    current_RouterConfiguration.id().c_str());
+
+      // check if current_router_subnet_id already exist in new_subnet_routing_tables
+      if (new_subnet_routing_tables.find(current_router_subnet_id) !=
+          new_subnet_routing_tables.end()) {
+        is_subnet_routing_table_exist = true;
+      }
+
+      // Look up the subnet configuration to query for additional info
+      for (int j = 0; j < parsed_struct.subnet_states_size(); j++) {
+        SubnetConfiguration current_SubnetConfiguration =
+                parsed_struct.subnet_states(j).configuration();
+
+        ACA_LOG_DEBUG("current_SubnetConfiguration subnet ID: %s.\n",
+                      current_SubnetConfiguration.id().c_str());
+
+        if ((parsed_struct.subnet_states(j).operation_type() == OperationType::INFO) &&
+            (current_SubnetConfiguration.id() == current_router_subnet_id)) {
+          found_vpc_id = current_SubnetConfiguration.vpc_id();
+
+          found_cidr = current_SubnetConfiguration.cidr();
+
+          slash_pos = found_cidr.find('/');
+          if (slash_pos == string::npos) {
+            throw std::invalid_argument("'/' not found in cidr");
+          }
+
+          found_tunnel_id = current_SubnetConfiguration.tunnel_id();
+          if (!aca_validate_tunnel_id(found_tunnel_id)) {
+            throw std::invalid_argument("found_tunnel_id is invalid");
+          }
+
+          // subnet info's gateway ip and mac needs to be there and valid
+          found_gateway_ip = current_SubnetConfiguration.gateway().ip_address();
+
+          // inet_pton returns 1 for success 0 for failure
+          if (inet_pton(AF_INET, found_gateway_ip.c_str(), &(sa.sin_addr)) != 1) {
```

Comment:
What is sa.sin_addr here?

Reply:
Should we wrap this validation as a util method?

Reply:
that's the 'c' way to validate IP address. Yes, we can switch to util method next time when we touch this.